### PR TITLE
Add foreign sub decl AST node type

### PR DESCRIPTION
**Pull request by rightfold**

### CODE REVIEW

File: src-lib/Language/Mill/AST.hs [hunk 1]
```diff
@@ -1,46 +1,34 @@
...
-, Type(..)
-, Module(..)
-, Decl(..)
-, Expr(..)
-, Stmt(..)
-) where
-
-data ModuleName = ModuleName [String]
-  deriving (Eq, Show)
+newtype ModuleName = ModuleName [String] deriving (Eq, Show)
```
Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: I'd rather we add newlines each time for a `deriving`

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: OK

File: src-lib/Language/Mill/AST.hs [hunk 2]
```diff
@@ -1,46 +1,40 @@
...
 data Decl
     = ImportDecl ModuleName
     | SubDecl String ParameterList Type Expr
-  deriving (Eq, Show)
+    | ForeignSubDecl ForeignSource CallingConvention String ParameterList Type
+    deriving (Eq, Show)
+
+newtype ForeignSource = ForeignSource String
+                        deriving (Eq, Show)
+newtype CallingConvention = CallingConvention Name
```
Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: shouldn't we constrain this? why is that a `Name`? Can I have `CallingConvention $ QualifiedName (ModuleName ["org", "rightfold"]) "hey"`? what would that mena?

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: Yes, but that'd be an error if `org.rightfold.hey` isn't a calling convention.

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: Calling conventions are imported from modules that are known to the compiler, e.g.

    import mill.foreign.ecmascript
    foreign "./a.js" sub ecmascript.returnCall f(): Unit

Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: reads a bit funny, no?

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: how come?

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: yeah the order is a bit weird. may have to change that

Comment (rightfold) [inline src-lib/Language/Mill/AST.hs]: putting the JS filename directly after `foreign` allows for declaration grouping:

    foreign "./a.js" (
        sub ecmascript.returnCall f(): Unit
        sub ecmascript.returnCall g(): Unit
    )

Comment (vendethiel) [inline src-lib/Language/Mill/AST.hs]: looks better :+1:

**Pull request merged**